tools/mkldscript.c

```diff
@@ -1,298 +1,465 @@
+        "{\n"
+        "\t_RomStart = 0;\n"
+        "\t_RomSize = 0;\n"
+        "\n"
+    );
+
+    for (i = 0; i < segment_count; i++)
+    {
+        const struct Segment *seg = &segments[i];
 
-        ////fprintf(fout, "        . = ALIGN(0x10);\n");
+        /* begin segment */
+        if (seg->fields & (1 << STMT_romalign))
+            fprintf(f, "\t_RomSize = ALIGN(_RomSize, %i);\n", seg->romalign);
 
-        //fprintf(fout, "        _%sSegmentOvlEnd = .;\n", seg->name);
+        fprintf(f,
+            "\t_%sSegmentRomStartTemp = _RomSize;\n"
+            "\t_%sSegmentRomStart = _%sSegmentRomStartTemp;\n",
+            seg->name,
+            seg->name, seg->name
+        );
 
-        //fprintf(fout, "\n    }\n");
+        /* placement */
+        fprintf(f, "\t..%s ", seg->name);
+        if (seg->fields & (1 << STMT_after))
+            fprintf(f, "_%sSegmentEnd ", seg->after);
+        else if (seg->fields & (1 << STMT_number))
+            fprintf(f, "0x%02X000000 ", seg->number);
+        else if (seg->fields & (1 << STMT_address))
+            fprintf(f, "0x%08X ", seg->address);
+        fprintf(f, ": AT(_%sSegmentRomStart)\n\t{\n", seg->name);
+    
+        /* content sections */
+        fprintf(f, "\t\t_%sSegmentStart = .;\n\n", seg->name);
+        fprintf(f, "\t\t%s/%s.o (.text)\n", dir, seg->name);
+        fprintf(f, "\t\t%s/%s.o (.data)\n", dir, seg->name);
+        fprintf(f, "\t\t%s/%s.o (.rodata)\n", dir, seg->name);
+
+        /* overlay relocation section */
+        fprintf(f, "\t\t_%sSegmentOvlStart = .;\n", seg->name);
+
+        if (strncmp(seg->name, "ovl_", 4) == 0)
+            fprintf(f, "\t\t%s/%s.reloc.o (.ovl)\n", dir, seg->name);
+
+        fprintf(f,
+            "\t\t_%sSegmentOvlEnd = .;\n"
+            "\t\t_%sSegmentOvlSize = ABSOLUTE(_%sSegmentOvlEnd - _%sSegmentOvlStart);\n",
+            seg->name,
+            seg->name, seg->name, seg->name
+        );
+
+        /* loadable sections done */
+        if (seg->fields & (1 << STMT_increment))
+            fprintf(f, "\n\t. += 0x%08X;\n", seg->increment);
+
+        fprintf(f,
+            "\t}\n"
+            "\t_%sSegmentRomSize = ABSOLUTE(_%sSegmentOvlEnd - _%sSegmentTextStart);\n"
+            "\t_RomSize += _%sSegmentRomSize;\n"
+            "\t_%sSegmentRomEndTemp = _RomSize;\n"
+            "\t_%sSegmentRomEnd = _%sSegmentRomEndTemp;\n",
+            seg->name, seg->name, seg->name,
+            seg->name,
+            seg->name,
+            seg->name, seg->name
+        );
+
+        if (seg->fields & (1 << STMT_romalign))
+            fprintf(f, "\t_RomSize = ALIGN(_RomSize, %i);\n", seg->romalign);
+
+        /* .bss sections */
+        fprintf(f,
+            "\t..%s.bss (NOLOAD) :\n"
+            "\t{\n"
+            "\t\t%s/%s.o (.bss)\n"
+            "\t}\n",
+            seg->name,
+            dir, seg->name
+        );
+
+        /* end segment */
+        fprintf(f,
+            "\t\t_%sSegmentEnd = .;\n"
```

Comment:
you have probably looked at it but just in case this is the makerom segment part of new `build/ldscript.txt` :

<details>

```
	_makeromSegmentRomStartTemp = _RomSize;
	_makeromSegmentRomStart = _makeromSegmentRomStartTemp;
	..makerom : AT(_makeromSegmentRomStart)
	{
		_makeromSegmentStart = .;

		build/segments/makerom.o (.text)
		build/segments/makerom.o (.data)
		build/segments/makerom.o (.rodata)
		_makeromSegmentOvlStart = .;
		_makeromSegmentOvlEnd = .;
		_makeromSegmentOvlSize = ABSOLUTE(_makeromSegmentOvlEnd - _makeromSegmentOvlStart);
	}
	_makeromSegmentRomSize = ABSOLUTE(_makeromSegmentOvlEnd - _makeromSegmentTextStart);
	_RomSize += _makeromSegmentRomSize;
	_makeromSegmentRomEndTemp = _RomSize;
	_makeromSegmentRomEnd = _makeromSegmentRomEndTemp;
	..makerom.bss (NOLOAD) :
	{
		build/segments/makerom.o (.bss)
	}
		_makeromSegmentEnd = .;
		_makeromSegmentSize = ABSOLUTE(_makeromSegmentEnd - _makeromSegmentStart);
```

</details>

So first congrats again because it's infinitely more legible than it was previously, but I must say the indent difference on `_makeromSegmentEnd` looks weird, unless that's intended to match `_makeromSegmentStart`?

Maybe you could `_makeromSegmentStart = ...` before `..makerom : AT(_makeromSegmentRomStart)`
and change `..makerom : AT(_makeromSegmentRomStart)` to `..makerom _makeromSegmentStart : AT(_makeromSegmentRomStart)`
if the linker script format allows it. (I'm not an expert)
Then `_makeromSegmentStart` and `_makeromSegmentEnd` would be on the same indentation level
(think of my OCD tharo 😢)